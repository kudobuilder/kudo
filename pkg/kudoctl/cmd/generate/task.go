package generate

import (
	"fmt"
	"path"

	"github.com/spf13/afero"

	"github.com/kudobuilder/kudo/pkg/apis/kudo/v1beta1"
	"github.com/kudobuilder/kudo/pkg/engine/task"
	"github.com/kudobuilder/kudo/pkg/kudoctl/packages/reader"
)

// AddTask adds a task to the operator.yaml file
func AddTask(fs afero.Fs, path string, task *v1beta1.Task) error {
	p, err := reader.ReadDir(fs, path)
	if err != nil {
		return err
	}
	o := p.Files.Operator

	o.Tasks = append(o.Tasks, *task)

	return writeOperator(fs, path, o)
}

// TaskList provides a list of operator tasks
func TaskList(fs afero.Fs, path string) ([]v1beta1.Task, error) {
	p, err := reader.ReadDir(fs, path)
	if err != nil {
		return nil, err
	}

	return p.Files.Operator.Tasks, nil
}

func TaskInList(fs afero.Fs, path, taskName string) (bool, error) {
	list, err := TaskList(fs, path)
	if err != nil {
		return false, err
	}
	for _, task := range list {
		if task.Name == taskName {
			return true, nil
		}
	}
	return false, nil
}

// TaskKinds provides a list of task kinds what are supported via generators
func TaskKinds() []string {
	return []string{task.ApplyTaskKind, task.DeleteTaskKind, task.PipeTaskKind, task.ToggleTaskKind}
}

// EnsureTaskResources ensures that all resources used by the given task exist
func EnsureTaskResources(fs afero.Fs, path string, task *v1beta1.Task) error {

	for _, resource := range task.Spec.Resources {
		err := EnsureResource(fs, path, resource)
		if err != nil {
			return err
		}
	}

	if task.Spec.Pod != "" {
		err := EnsureResource(fs, path, task.Spec.Pod)
		if err != nil {
			return err
		}
	}
	return nil
}

// EnsureResource ensures that resource is in templates folder
func EnsureResource(fs afero.Fs, dir string, resource string) error {

	// does "operator" path exist?  if not err
	exists, err := afero.DirExists(fs, dir)
	if err != nil {
		return err
	}
	if !exists {
		return fmt.Errorf("operator path %q does not exist", dir)
	}

	// does templates folder exist? if not mkdir
	templatePath := path.Join(dir, "templates")
	exists, err = afero.DirExists(fs, templatePath)
	if err != nil {
		return err
	}
	if !exists {
		err = fs.Mkdir(templatePath, 0755)
		if err != nil {
			return err
		}
	}

	// does resource exist? if not "touch" it, otherwise good
	resourcePath := path.Join(dir, "templates", resource)
	exists, err = afero.Exists(fs, resourcePath)
	if err != nil {
		return err
	}
	if !exists {
		err = afero.WriteFile(fs, resourcePath, []byte{}, 0755)
		if err != nil {
			return err
		}
	}
	return nil
}
