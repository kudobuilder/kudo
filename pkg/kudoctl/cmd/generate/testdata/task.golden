apiVersion: kudo.dev/v1beta1
appVersion: 3.4.10
kubernetesVersion: 1.15.0
kudoVersion: 0.2.0
maintainers:
- email: avarkockova@mesosphere.com
  name: Alena Varkockova
- email: runyontr@gmail.com
  name: Tom Runyon
- email: kensipe@gmail.com
  name: Ken Sipe
name: zookeeper
operatorVersion: 0.1.0
plans:
  deploy:
    phases:
    - name: zookeeper
      steps:
      - name: everything
        tasks:
        - infra
        - app
      strategy: parallel
    strategy: serial
  validation:
    phases:
    - name: connection
      steps:
      - name: connection
        tasks:
        - validation
      strategy: parallel
    strategy: serial
tasks:
- kind: Apply
  name: infra
  spec:
    resources:
    - services.yaml
    - pdb.yaml
- kind: Apply
  name: app
  spec:
    resources:
    - statefulset.yaml
- kind: Apply
  name: validation
  spec:
    resources:
    - validation.yaml
- kind: Apply
  name: Foo
  spec:
    resources:
    - bar.yaml
url: https://zookeeper.apache.org/
